vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,787 @@
+  std::vector<std::vector<VW::namespace_index>> base_interactions;
+
+  ex->interactions = &base_interactions;
+  auto* ex1 = VW::flatten_sort_example(all, ex);
+  for (auto& f : ex1->fs) { base.emplace_back(f.index(), f.value()); }
+  VW::free_flatten_example(ex1);
+
+  ex->interactions = full_interactions;
+  auto* ex2 = VW::flatten_sort_example(all, ex);
+  for (auto& f : ex2->fs) { full.emplace_back(f.index(), f.value()); }
+  VW::free_flatten_example(ex2);
+}
+
+emt_lru::emt_lru(unsigned long max_size) { (*this).max_size = max_size; }
+
+emt_lru::K emt_lru::bound(emt_lru::K item)
+{
+  if (max_size == 0) { return nullptr; }
+
+  auto item_map_reference = map.find(item);
+
+  // item is not in map so we add it to our list
+  if (item_map_reference == map.end())
+  {
+    list.push_front(item);
+    map.emplace(std::pair<K, V>(item, list.begin()));
+  }
+  // item is in map already so we move it to the front of the line
+  else
+  {
+    V item_list_reference = (*item_map_reference).second;
+    if (item_list_reference != list.begin())
+    {
+      list.splice(list.begin(), list, item_list_reference, std::next(item_list_reference));
+    }
+  }
+
+  if (list.size() > max_size)
+  {
+    K last_value = list.back();
+    list.pop_back();
+    map.erase(last_value);
+    return last_value;
+  }
+  else { return nullptr; }
+}
+
+emt_node::emt_node()
+{
+  left = nullptr;
+  right = nullptr;
+  router_decision = 0;
+}
+
+emt_tree::emt_tree()
+{
+  leaf_split = 100;
+
+  scorer_type = emt_scorer_type::self_consistent_rank;
+  router_type = emt_router_type::eigen;
+
+  begin = 0;
+
+  all = nullptr;
+  ex = nullptr;
+  bounder = nullptr;
+  root = VW::make_unique<emt_node>();
+
+  // we set this up for repeated use later in the scorer.
+  // we will populate this examples features over and over.
+  ex = VW::make_unique<VW::example>();
+  ex->interactions = new std::vector<std::vector<VW::namespace_index>>();
+  ex->extent_interactions = new std::vector<std::vector<extent_term>>();
+  ex->indices.push_back(0);
+}
+
+emt_tree::~emt_tree()
+{
+  delete ex->interactions;
+  delete ex->extent_interactions;
+}
+
+////////////////////////end of definitions/////////////////
+///////////////////////////////////////////////////////////
+
+///////////////////////Helper/////////////////////////////
+//////////////////////////////////////////////////////////
+float emt_median(std::vector<float>& array)
+{
+  int size = array.size();
+  auto nth = array.begin() + size / 2;
+
+  if (size % 2 == 0)
+  {
+    std::nth_element(array.begin(), nth, array.end());
+    auto v1 = (*nth);
+    std::nth_element(array.begin(), nth - 1, array.end());
+    auto v2 = *(nth - 1);
+    return (v1 + v2) / 2.;
+  }
+  else
+  {
+    std::nth_element(array.begin(), nth, array.end());
+    return *nth;
+  }
+}
+
+float emt_inner(const emt_feats& xs, const emt_feats& ys)
+{
+  float sum = 0;
+  unsigned long xi = 0;
+  unsigned long yi = 0;
```

Comment:
Prefer explicit sized types, unless it needs to be `size_t` which looks like it may be the case.